### PR TITLE
samples: usb: add UAC2 implicit feedback sample

### PR DESCRIPTION
**Pull request by tmon-nordic**
The sample illustates how to achieve bidirectional asynchronous audio. Implicit feedback is the only way to achieve asynchronous USB headset on devices that have only one isochronous IN and one isochronous OUT endpoint.

The sample implements stereo playback and mono recording. While it would be possible to have stereo recording, the commonly available headsets are mono only. The number of channels can differ between audio sink and source because the only requirement is that the two run on same clock.

### CODE REVIEW

File: samples/subsys/usb/uac2_implicit_feedback/prj.conf
Comment (koffes) [inline samples/subsys/usb/uac2_implicit_feedback/prj.conf:16]: Do we need to use a NRFX type timer here, or can a generic Zephyr timer be used?

File: samples/subsys/usb/uac2_implicit_feedback/src/feedback_dummy.c
Comment (koffes) [inline samples/subsys/usb/uac2_implicit_feedback/src/feedback_dummy.c:17]: With should normally be executed in these functions? Can some comments be added to guide customers?

File: samples/subsys/usb/uac2_implicit_feedback/src/main.c
Comment (koffes) [inline samples/subsys/usb/uac2_implicit_feedback/src/main.c:46]: Would it be more descriptive to name this playback and recording? Or out and in?

File: samples/subsys/usb/uac2_implicit_feedback/prj.conf
Comment (tmon-nordic) [reply samples/subsys/usb/uac2_implicit_feedback/prj.conf:16]: `feedback_nrf53.c` implementation uses timer and connects USB SOF and I2S FRAMESTART to it with DPPI. DPPI is so specific to Nordic that I don't really see how to achieve this without resorting to low-level hardware interface.

I don't know how to achieve proper feedback without relying on the low-level functionality. This however is something that application has to somehow figure out and I do provide a working sample implementation for nRF53.

File: samples/subsys/usb/uac2_implicit_feedback/src/feedback_dummy.c
Comment (tmon-nordic) [reply samples/subsys/usb/uac2_implicit_feedback/src/feedback_dummy.c:17]: Read the comment in feedback.h:
https://github.com/zephyrproject-rtos/zephyr/blob/602054402313f885c2a722616ceee87c735668b6/samples/subsys/usb/uac2_implicit_feedback/src/feedback.h#L22-L33

The comment refers to what has to be reported. How this is achieved is entirely up to the implementer.

I only know how to achieve it using DPPI subsystem and TIMER on nRF53. I hope that eventually someone will come up with platform independent solution, but this is not a trivial task at all and I think it would require some sort of interrupt/event timestamping API that is simply available yet (and some clever filtering because the control goal is to have I2S FRAMESTART and USB SOF happen at the same time, where obviously one interrupt handler will execute before the other and the two will keep "swapping" with each other which one executes first).

**Comment by carlescufi** [comment]
@koffes please take another look

File: samples/subsys/usb/uac2_implicit_feedback/src/main.c
Comment (tmon-nordic) [reply samples/subsys/usb/uac2_implicit_feedback/src/main.c:55]: There are 5 `bool` variables before this `uint8_t`, so moving this `uint8_t` won't really save anything.

On nRF5340DK the `sizeof(struct usb_i2s_ctx)` is 24 bytes. When `i2s_counter` is moved to the end of structure the size remains unchanged at 24 bytes. Moving all pointers to the beginning also makes the size remain intact.

Because there are 9 single-byte members (5 `bool` and 4 `uint8_t`) separated into two "blocks", shuffling them around won't give any structure size reduction regardless if the pointer size is 4 or 8 bytes.

File: samples/subsys/usb/uac2_implicit_feedback/prj.conf
Comment (tmon-nordic) [reply samples/subsys/usb/uac2_implicit_feedback/prj.conf]: Yes, moved.

**Review by jfischer-no — APPROVED** [review]

File: samples/subsys/usb/uac2_implicit_feedback/src/main.c
Comment (jfischer-no) [reply samples/subsys/usb/uac2_implicit_feedback/src/main.c:55]: Indeed, bool, "the default is CHAR_TYPE_SIZE"

File: samples/subsys/usb/uac2_implicit_feedback/sample.yaml
Comment (jfischer-no) [inline samples/subsys/usb/uac2_implicit_feedback/sample.yaml:6]: Must be `     - usbd`

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Review by jfischer-no — DISMISSED** [review]

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Review by jfischer-no — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
